### PR TITLE
[idea][aa] URIHandler based implementation of FileSystemSupport

### PR DESCRIPTION
**Pull request by akosyakov**

### CODE REVIEW

**Comment by akosyakov** [comment]
Jenkins intellij-tests-master-latest

**Comment by akosyakov** [comment]
Jenkins intellij-tests-master-latest

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/resource/IdeaResourceSetProvider.xtend [hunk 1]
```diff
@@ -155,15 +168,81 @@ class IdeaResourceSetProvider {
...
+				return attributes
+			}
+
+			val file = uri.virtualFile
+			if (file != null) {
+				val attributes = newHashMap
+				if (requestedAttributes.contains(ATTRIBUTE_DIRECTORY))
+					attributes.put(ATTRIBUTE_DIRECTORY, file.directory)
+				if (requestedAttributes.contains(ATTRIBUTE_TIME_STAMP))
+					attributes.put(ATTRIBUTE_TIME_STAMP, file.modificationStamp)
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/resource/IdeaResourceSetProvider.xtend]: should be file.timestamp

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/trace/VirtualFileBasedTrace.xtend [hunk 2]
```diff
@@ -178,7 +184,8 @@ class VirtualFileBasedTrace extends AbstractTrace implements IIdeaTrace {
 	}
 
 	override IdeaModuleConfig getLocalProjectConfig() {
-		return new IdeaModuleConfig(getLocalProject());
+		val module = localProject
+		return workspaceConfigProvider.getWorkspaceConfig(module).findProjectByName(module.name)
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/trace/VirtualFileBasedTrace.xtend]: Why is this necessary? Looks like unnecessary overhead to me at a first glance

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend [hunk 3]
```diff
@@ -56,5 +69,220 @@ abstract class AbstractFileSystemSupport implements MutableFileSystemSupport {
...
+		
+		try {
+			val out = URIConverter.createOutputStream(uri)
+			try {
+				ByteStreams.copy(source, out)
+			} finally {
+				out.close
+			}
+		} catch (IOException exc) {
+			throw new IllegalArgumentException(exc.message, exc)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend]: Maybe we should document the thrown exceptions and failure conditions also on the intf

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend [hunk 4]
```diff
@@ -56,5 +69,220 @@ abstract class AbstractFileSystemSupport implements MutableFileSystemSupport {
...
+	
+	override getContentsAsStream(Path path) {
+		val uri = path.URI
+		if (uri == null || !uri.exists || !isFile(uri))
+			throw new IllegalAccessException('The file cannot be found: ' + path)
+		
+		try {
+			return URIConverter.createInputStream(uri)
+		} catch (IOException exc) {
+			throw new IllegalArgumentException(exc.message, exc)
```
Comment (szarnekow) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend]: dito

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend [hunk 5]
```diff
@@ -1,24 +1,34 @@
...
 import org.eclipse.xtext.util.StringInputStream
-import org.eclipse.emf.ecore.resource.ResourceSet
+import org.eclipse.xtext.workspace.IWorkspaceConfigProvider
+
+import static org.eclipse.emf.ecore.resource.URIConverter.*
+import static org.eclipse.xtend.lib.macro.file.Path.*
+
+import static extension org.eclipse.xtext.util.UriUtil.*
 
 abstract class AbstractFileSystemSupport implements MutableFileSystemSupport {
```
Comment (szarnekow) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend]: There's a lot of sneaky throw in the generated code for this class. The failure cases should all be handled / wrapped properly and documented on the API.

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend [hunk 6]
```diff
@@ -1,24 +1,34 @@
...
 
 	@Inject @Accessors IEncodingProvider encodingProvider
 	@Inject(optional=true) @Accessors IFilePostProcessor postProcessor
+	@Inject @Accessors extension IWorkspaceConfigProvider workspaceConfigProvider
 	@Accessors ResourceSet context
```
Comment (szarnekow) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend]: This looks a little strange to me. Can we use an URIConverter directly?

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/filesystem/IdeaWorkspaceConfigProvider.xtend [hunk 7]
```diff
@@ -46,50 +56,56 @@ class IdeaWorkspaceConfig implements IWorkspaceConfig {
...
+			
+		return new IdeaModuleConfig(module, contentRoot)
 	}
-
-
-	private def isUsable(Module module) {
-		!ModuleRootManager.getInstance(module).contentEntries.isEmpty
+	
+	override getPath() {
+		URI.createURI(project.baseDir.url).toFolderURI
```
Comment (szarnekow) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/filesystem/IdeaWorkspaceConfigProvider.xtend]: I don't understand why this is a meaningful information in idea? The project file can be put anywhere and is therefore unrelated to every known virtual file, isn't it?

File: tests/org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/macro/JavaIoFileSystemTest.xtend [hunk 8]
```diff
@@ -45,95 +46,104 @@ class JavaIoFileSystemTest {
...
-		
+
 		path.delete
 		assertFalse(path.exists)
 		assertFalse(path.isFile)
 		assertFalse(path.isFolder)
 	}
-	
+
 	@Test def void testModificationStamp() {
```
Comment (szarnekow) [inline tests/org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/macro/JavaIoFileSystemTest.xtend]: We need a second test that ensures that the modification stamp is incremented

File: intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/trace/VirtualFileBasedTrace.xtend (hunk 2 above)
Comment (akosyakov) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/trace/VirtualFileBasedTrace.xtend]: Could we IdeaModuleConfig as a localProject for VirtualFileBasedTrace? Only client now is org.eclipse.xtext.idea.trace.TraceForVirtualFileProvider.newAbstractTrace(VirtualFileInProject) which actually already has an instance of IdeaModuleConfig.

Comment (akosyakov) [inline intellij/org.eclipse.xtext.idea/src/org/eclipse/xtext/idea/trace/VirtualFileBasedTrace.xtend]: Using the first content root is bogus, because a file is not necessary belongs to it. We should use org.eclipse.xtext.idea.filesystem.IdeaWorkspaceConfig.findProjectContaining(URI) instead.

**Comment by akosyakov** [comment]
Jenkins intellij-tests-master-latest

**Comment by akosyakov** [comment]
Jenkins intellij-tests-master-latest
